packages/react-components/react-combobox/src/ComboboxBase/useComboboxBaseSlots.ts

```diff
@@ -0,0 +1,138 @@
+import * as React from 'react';
+import { useMergedRefs } from '@fluentui/react-utilities';
+import { getDropdownActionFromKey, getIndexFromAction } from '../utils/dropdownKeyActions';
+import type { ComboboxState } from '../components/Combobox/Combobox.types';
+import type { DropdownState } from '../components/Dropdown/Dropdown.types';
+import type { ComboboxBaseProps } from './ComboboxBase.types';
+
+export const useComboboxBaseSlots = (props: ComboboxBaseProps, state: ComboboxState | DropdownState) => {
+  const { multiselect, placeholder } = props;
+  const {
+    activeOption,
+    getCount,
+    getIndexOfId,
+    getOptionAtIndex,
+    open,
+    selectOption,
+    setActiveOption,
+    setOpen,
+    value,
+  } = state;
+
+  // handle trigger focus/blur
+  const triggerRef = React.useRef<HTMLButtonElement | HTMLInputElement>(null);
+  const ignoreTriggerBlur = React.useRef(false);
+
+  /* Add listbox props and event handlers  */
+  state.listbox = {
+    multiselect,
+    tabIndex: undefined,
+    ...state.listbox,
+  };
+
+  /*
+   * Handle focus when clicking the listbox popup:
+   * 1. Move focus back to the button/input when the listbox is clicked (otherwise it goes to body)
+   * 2. Do not close the listbox on button/input blur when clicking into the listbox
+   */
+  const { onClick: onListboxClick, onMouseDown: onListboxMouseDown } = state.listbox;
+  state.listbox.onClick = (event: React.MouseEvent<HTMLDivElement>) => {
+    triggerRef.current?.focus();
+
+    onListboxClick?.(event);
+  };
+
+  state.listbox.onMouseDown = (event: React.MouseEvent<HTMLDivElement>) => {
+    ignoreTriggerBlur.current = true;
+
+    onListboxMouseDown?.(event);
+  };
+
+  /* Add trigger props and event handlers */
+  let triggerSlot;
+
+  if ((state as ComboboxState).input) {
```

Comment:
All of these `as` casts are not ideal. Does it work to do this instead?
```suggestion
  if ('input' in state) {
```

That would (hopefully) let typescript figure out that the `state` object is ComboboxState, and let you avoid the `as` casts both in the `if` and `else` clauses.

More info: https://css-tricks.com/typescript-discriminated-unions/#aa-narrowing

Reply:
I went back and forth on this a bit -- I could return a props object for the trigger and listbox and the event handlers directly instead of applying them to the slot, but it felt much less neat in practice. Do you have any strong opinions?

Another option is something simliar to what `useARIAButton` is doing, where the hook gets passed the shorthand props and options, and then calls `resolveShorthand` in the hook. The only issue there is how to get the proper typing for the ref, I think.

Reply:
You could let the individual `useCombobox`/`useDropdown` hooks handle the default props for these slots (basically everything except `ref` I think?)

Then you could take a param `triggerSlot: ComboboxState['input'] | DropdownState['button']`. You might need to cast triggerRef to `React.RefObject<unknown>` or `<never>` to work here.

Reply:
You could also skip taking it as a param and initialize it as:
```ts
const triggerSlot = ('input' in state) ? state.input : state.button;
```